src/options_description.cpp

```diff
@@ -122,30 +126,35 @@ namespace boost { namespace program_options {
 
     const std::string& 
     option_description::key(const std::string& option) const
-    {        
-        if (!m_long_name.empty()) 
-            if (m_long_name.find('*') != string::npos)
+    {
+        // We make the arbitrary choise of using the first long
+        // name as the key, regardless of anything else
+        if (!m_long_names.empty()) {
+            const std::string& first_long_name = *m_long_names.begin();
+            if (first_long_name.find('*') != string::npos)
```

Comment:
I am not sure this logic is correct. If we have 3 longs options, with the second one matching the input and having "*" at the end, we want to return the actual matched token? This logic will return the matched token only if the first option has trailing "*"

Reply:
I don't quite get your example. What do you mean by 

> "" at the end

? End of what, and what's at that end?

Also, remember that this is `key()`, not `match()`; in `match()` we're obviously more lenient.

Reply:
By the way... is program_options supposed to support an option having an empty long name and no short name? I don't see code in the current `set_name()` which specifically rules this out (e.g. an `assert()`).

Reply:
This code handles a case where long option name has an asterisk at the end (looks like github helpfully removed those asterisks in my original comment. Say, we have long name of `foo*`, it will match either `foo1` or `foo2`, and the `key` method will notice his case and use `foo1` or `foo2` respectively, not `foo*`. As far as I can see, with your patch, given `bar,foo*` as long options, and `foo1` as token, the method will return `bar`, not `foo1`.

Reply:
Empty long name and no short name is clearly not a useful case, so a diagnostic might be helpful.